File: app/src/main/java/com/bintonet/android/kotlindashboard/customviews/ProgressCircle.java
package com.bintonet.android.kotlindashboard.customviews;

import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.SweepGradient;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.AccelerateDecelerateInterpolator;

import com.bintonet.android.kotlindashboard.R;

/**
 * CustomView for drawing the score in the form of a progress ring
 */

public class ProgressCircle extends View {

    private final RectF mOval = new RectF();
    private float mSweepAngle = 0;
    private int startAngle = 90;

    float mEndAngle = 1.0f;

    float shaderCx, shaderCy;
    int shaderColor0 = Color.YELLOW;
    int shaderColor1 = Color.RED;

    Paint progressPaint = new Paint();

    Paint incompletePaint = new Paint();

    private float strokeWidth = 4.0f;

    public ProgressCircle(Context context, AttributeSet attrs) {
        super(context, attrs);

        TypedArray a = context.getTheme().obtainStyledAttributes(
                attrs,
                R.styleable.ProgressCircle,
                0, 0);

        int incompleteColor;

        try {

            strokeWidth = a.getDimension(R.styleable.ProgressCircle_strokeWidth, 15.0f);
            incompleteColor = a.getColor(R.styleable.ProgressCircle_incompleteProgressColor, 0xff000000);
        } finally {
            a.recycle();
        }

        progressPaint.setStrokeWidth(strokeWidth);
        progressPaint.setStyle(Paint.Style.STROKE);
        progressPaint.setFlags(Paint.ANTI_ALIAS_FLAG);
        progressPaint.setShader(new SweepGradient(shaderCx, shaderCy, shaderColor0, shaderColor1));

        incompletePaint.setColor(incompleteColor);
        incompletePaint.setStrokeWidth(strokeWidth);
        incompletePaint.setStyle(Paint.Style.STROKE);
        incompletePaint.setFlags(Paint.ANTI_ALIAS_FLAG);

    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        float w, h;

        w = getWidth();
        h = getHeight();
        shaderCx = w/2;
        shaderCy = h/2;

        mOval.set(strokeWidth / 2, strokeWidth / 2, w - (strokeWidth / 2), w - (strokeWidth / 2));
        canvas.drawArc(mOval, -startAngle, (mSweepAngle * 360), false,
                progressPaint);

        mOval.set(strokeWidth / 2, strokeWidth / 2, w - (strokeWidth / 2), w - (strokeWidth / 2));
        canvas.drawArc(mOval, mSweepAngle * 360- startAngle, 360 - (mSweepAngle * 360), false,
                incompletePaint);

    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, widthMeasureSpec);
    }


    public void setProgress(float progress) {
        if (progress > 1.0f || progress < 0) {
            throw new RuntimeException("Value must be between 0 and 1: " + progress);
        }

        mEndAngle = progress;

        this.invalidate();
    }

    public void startAnimation() {
        ValueAnimator anim = ValueAnimator.ofFloat(mSweepAngle, mEndAngle);
        anim.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator valueAnimator) {
                ProgressCircle.this.mSweepAngle = (Float) valueAnimator.getAnimatedValue();
                ProgressCircle.this.invalidate();
            }
        });
        anim.setDuration(500);
        anim.setInterpolator(new AccelerateDecelerateInterpolator());
        anim.start();

    }

}

File: app/src/main/java/com/bintonet/android/kotlindashboard/customviews/ProgressCircle.kt
package com.bintonet.android.kotlindashboard.customviews

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.SweepGradient
import android.util.AttributeSet
import android.view.View
import android.view.animation.AccelerateDecelerateInterpolator

import com.bintonet.android.kotlindashboard.R

/**
 * CustomView for drawing the circle for the score
 */

class ProgressCircle(context: Context, attrs: AttributeSet) : View(context, attrs) {

    private val mOval = RectF()
    private var mSweepAngle = 0f
    private var mEndAngle = 1.0f

    private var shaderCx: Float = 0.toFloat()
    private var shaderCy: Float = 0.toFloat()
    private var shaderColor0 = Color.YELLOW
    private var shaderColor1 = Color.RED

    private var progressPaint = Paint()

    private var strokeWidth = 4.0f

    init {

        // here we load any custom attrs set in the layout xml
        val a = context.theme.obtainStyledAttributes(
                attrs,
                R.styleable.ProgressCircle,
                0, 0)

        try {
            //we store them here as variables for use later
            strokeWidth = a.getDimension(R.styleable.ProgressCircle_strokeWidth, 4.0f)
        } finally {
            a.recycle()
        }

        // building the paint object for the circle
        progressPaint.strokeWidth = strokeWidth
        progressPaint.style = Paint.Style.STROKE
        progressPaint.flags = Paint.ANTI_ALIAS_FLAG
        progressPaint.shader = SweepGradient(shaderCx, shaderCy, shaderColor0, shaderColor1)

    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        // set some dimensions for the circle (width and height of the canvas)
        val w: Float = width.toFloat()
        val h: Float = height.toFloat()

        //these are used for the colour gradient
        shaderCx = w / 2
        shaderCy = h / 2

        //now we set the Rectf coordinates
        mOval.set(strokeWidth / 2, strokeWidth / 2, w - strokeWidth / 2, w - strokeWidth / 2)

        //draw the circle now
        // startAngle : I am not sure why but 0 is at the right of the screen, so 270 is starting at the top
        // sweepangle, how far round we want the arc to go
        // usecenter, do we want the arc to go to the centre of the circle, if yes then the result would look like a pie chart
        // finally the paint we want to use, in our case the progress paint we defined earlier
        canvas.drawArc(mOval, 270F, mSweepAngle * 360, false, progressPaint)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, widthMeasureSpec)
    }


    // function we call from the mainview to set how far round the arc should go
    fun setProgress(progress: Float) {
        if (progress > 1.0f || progress < 0) {
            throw RuntimeException("Value must be between 0 and 1: " + progress)
        }
        mEndAngle = progress

        this.invalidate()
    }

    // simple animation function to be called after we set the progress to redraw the arc
    //  but have it draw over a set amount of time
    fun startAnimation() {
        val anim = ValueAnimator.ofFloat(mSweepAngle, mEndAngle)
        anim.addUpdateListener { valueAnimator ->
            this@ProgressCircle.mSweepAngle = valueAnimator.animatedValue as Float
            this@ProgressCircle.invalidate()
        }
        anim.duration = 300
        anim.interpolator = AccelerateDecelerateInterpolator()
        anim.start()

    }

}

